### PR TITLE
Implement metaclass based hierarchical drivers

### PR DESCRIPTION
**Pull request by MatthieuDartiailh**
This is a cleaned up and tested implementation of my proposal for hierarchical drivers. I tried to trim Eapii heritage but however everything related to caching and the actual implementation of the Features and how they hook up into the drivers are not definitive and still carry the weight of the heritage. This is rather to have a starting point but I chose to do it that way so that we can start testing.
As I don't have Python 3 running I tested only on Python 2 (I expect to get a new computer soon and will then set up a dual version system).

### CODE REVIEW

File: lantz_core/has_features.py [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+        # Create the class object.
+        cls = super(HasFeaturesMeta, meta).__new__(meta, name, bases, dct)
+
+        # Purge the list of base classes (for the most base class object is
+        # present)
+        bases = [b for b in bases if issubclass(b, AbstractHasFeatures)]
+
+        # Analyse the source code to find the doc for the defined Features.
+        # This will work as long as two subpart are not aliased in the same
+        # way which is probabbly good enough.
```
Comment (alexforencich) [inline lantz_core/has_features.py]: Interesting method for building the docstrings.  How is the performance of this method?

Comment (MatthieuDartiailh) [inline lantz_core/has_features.py]: I have never timed it. It happens only once when the class is created so it should not be too bad as long as people don't load the whole library of drivers at once. I like it as it is the natural way to document things in Sphinx so there is no surprise.

**Comment by MatthieuDartiailh** [comment]
@hgrecco comments addressed.

File: lantz_core/has_features.py (hunk 1 above)
Comment (hgrecco) [inline lantz_core/has_features.py]: I have never done this but it looks clean at indeed is only done once. Making it sphinx readable is a +1 for me

**Comment by MatthieuDartiailh** [comment]
Is everybody happy with that ? @LabPy/owners 
Any comments ? @p3trus, @crazyfermions

**Comment by p3trus** [comment]
Just a short comment: Maybe it's a good idea to hide BaseDriver methods e.g. with a single underscore to avoid name clashes with Features. E.g. some devices support a initialize commands which would shadow the one of the baseclass.

**Comment by MatthieuDartiailh** [comment]
True @p3trus. However as we expect users to open the connection themselves (@hgrecco was against the opening of the connection on driver \__init\__) they need to call it so we should rather rename it perhaps. @hgrecco you were the one in favor of having initialize rather than open_connection, should we change back (or find something else).

**Comment by hgrecco** [comment]
I think we should have an initialize and finalize method which can overridden by derived classes. Is the responsability of derived classes to call the parents class initialize and finalize.

**Comment by MatthieuDartiailh** [comment]
I can live with that, my only concern is that in some cases the user might want to open the connection but not call the initialize method of the instrument which set it in a kind on default mode (but in that case it should not be called initialized in the first place). So I say we can keep it.

**Comment by MatthieuDartiailh** [comment]
This is good to merge from my end. I think it is a good idea to have the reviewers rather than the proposer of a PR to merge so if any of you feels like pushing the green button he is welcome.

**Pull request merged**